### PR TITLE
Add parameter usage to CTC inference tutorial

### PR DESCRIPTION
**Pull request by carolineechen**
Add explanation and demonstration of different beam search decoder parameters.
Additionally use a better sample audio file and load in with token list instead of tokens file.

### CODE REVIEW

File: examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py [hunk 1]
```diff
@@ -161,26 +162,27 @@
...
-# :py:func:`torchaudio.prototype.ctc_decoder.kenlm_lexicon_decoder`
-# factory function.
-# In addition to the previously mentioned components, it also takes in
-# various beam search decoding parameters and token/word parameters.
+# The decoder can be constructed using the ``kenlm_lexicon_decoder``
+# factory function from ``torchaudio.prototype.ctc_decoder``. In addition
+# to the previously mentioned components, it also takes in various beam
+# search decoding parameters and token/word parameters. The full list of
+# parameters can be found
+# `here <https://pytorch.org/audio/main/prototype.html#kenlm-lexicon-decoder>`__.
```
Comment (mthrok) [inline examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py:170]:
```suggestion
# :py:func:`here <torchaudio.prototype. kenlm_lexicon_decoder>`.
```

Since once this tutorial is released, then we will have multiple versions of this tutorials, so I think it's better to use interlink.

File: examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py [hunk 2]
```diff
@@ -222,28 +224,212 @@ def forward(self, emission: torch.Tensor) -> str:
 #
 # Now that we have the data, acoustic model, and decoder, we can perform
 # inference. Recall the transcript corresponding to the waveform is
-# ``"when it was the seven hundred and eighteenth night"``
+# ``"i really was very much afraid of showing him how much shocked i was at some parts of what he said"``
```
Comment (mthrok) [inline examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py:227]: I think the code-block highlights this better. The following is the comparison of double back-ticks, single back-tick and code block.

```

``"i really was very much afraid of showing him how much shocked i was at some parts of what he said"``

`"i really was very much afraid of showing him how much shocked i was at some parts of what he said"`

.. code::

    i really was very much afraid of showing him how much shocked i was at some parts of what he sai\
d
```

Screen Shot 2022-01-05 at 1 36 32 PM

File: examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py [hunk 3]
```diff
@@ -222,28 +224,212 @@ def forward(self, emission: torch.Tensor) -> str:
...
 #
 
+actual_transcript = "i really was very much afraid of showing him how much shocked i was at some parts of what he said"
```
Comment (mthrok) [inline examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py:230]: This variable is always used as `actual_transcript.split()`, so I think you can `split()` here.

File: examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py [hunk 4]
```diff
@@ -222,28 +224,212 @@ def forward(self, emission: torch.Tensor) -> str:
...
 emission, _ = acoustic_model(waveform)
 
+
+######################################################################
+# The greedy decoder give the following result.
+#
+
+greedy_result = greedy_decoder(emission[0])
+greedy_transcript = greedy_result.replace("|", " ").lower().strip()
```
Comment (mthrok) [inline examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py:239]: I think `.replace("|", " ").lower().strip()` and `.split()` applied later can be put inside of `GreedyDecoder.forward` method.

File: examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py [hunk 5]
```diff
@@ -222,28 +224,212 @@ def forward(self, emission: torch.Tensor) -> str:
...
+print(f"Transcript: {beam_search_transcript}")
+print(f"WER: {beam_search_wer}")
+
 
 ######################################################################
-# Using the greedy decoder:
+# We see that the transcript with the lexicon-constrained beam search
+# decoder produces a more accurate result consisting of real words, while
+# the greedy decoder can predict incorrectly spelled words like “affrayd”
+# and “shoktd”.
```
Comment (mthrok) [inline examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py:266]: Nice!

File: examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py [hunk 6]
```diff
@@ -222,28 +224,212 @@ def forward(self, emission: torch.Tensor) -> str:
...
-print(greedy_transcript)
+
+######################################################################
+# Beam Search Decoder Parameters
+# ------------------------------
+#
+# In this section, we go a little bit more in depth about some different
+# parameters and tradeoffs. For the full list of customizable parameters,
+# please refer to the
+# `documentation <https://pytorch.org/audio/main/prototype.ctc_decoder.html#torchaudio.prototype.ctc_decoder.kenlm_lexicon_decoder>`__.  # noqa
```
Comment (mthrok) [inline examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py:277]: Similar to the previous one, use of `:py:func:` should is more future-proof.

File: examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py [hunk 7]
```diff
@@ -222,28 +224,212 @@ def forward(self, emission: torch.Tensor) -> str:
...
+# and “shoktd”.
+#
 
-greedy_result = greedy_decoder(emission[0])
-greedy_transcript = greedy_result.replace("|", " ").lower().strip()
-print(greedy_transcript)
+
+######################################################################
+# Beam Search Decoder Parameters
+# ------------------------------
```
Comment (mthrok) [inline examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py:272]: not the scope of this PR but can we show how scores are computed? 

Like the one from [this](https://docs.google.com/presentation/d/1y9IHcy8F8rywgoTyvQbcsnlzzMbCSbPT2rW58pKGxms/edit) or cite a publication if there is as the linked one is more on the side of the implementation detail.

File: examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py [hunk 8]
```diff
@@ -222,28 +224,212 @@ def forward(self, emission: torch.Tensor) -> str:
...
+
+
+######################################################################
+# additional parameters
+# ~~~~~~~~~~~~~~~~~~~~~
+#
+# Additional parameters that can be optimized include the following -
+# ``word_score``: score to add when word finishes - ``unk_score``: unknown
+# word appearance score to add - ``sil_score``: silence appearance score
+# to add - ``log_add``: whether to use log add for lexicon Trie smearing
```
Comment (mthrok) [inline examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py:434]: Should this be list layout?

File: examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py [hunk 9]
```diff
@@ -36,7 +36,7 @@
 # working with
 #
 
-import os
+from time import time
```
Comment (mthrok) [inline examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py:39]: Since it's used for measuring the execution time, should use [`monotonic`](https://docs.python.org/3/library/time.html#time.monotonic).

File: examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py (hunk 7 above)
Comment (carolineechen) [reply examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py:277]: sounds good, will leave this as a followup

File: examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py [hunk 10]
```diff
@@ -210,10 +213,11 @@ def forward(self, emission: torch.Tensor) -> str:
         indices = torch.argmax(emission, dim=-1)  # [num_seq,]
         indices = torch.unique_consecutive(indices, dim=-1)
         indices = [i for i in indices if i != self.blank]
-        return "".join([self.labels[i] for i in indices])
+        joined = "".join([self.labels[i] for i in indices])
+        return joined.replace("|", " ").lower().strip().split()
```
Comment (mthrok) [inline examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py:217]: Looking at `replace("|", " ")`, I wonder if we should add an argument to [`get_labels()`](https://pytorch.org/audio/main/pipelines.html#torchaudio.pipelines.Wav2Vec2ASRBundle.get_labels) to customize this. I did not add this because I was looking the labels from written language perspective and not all the languages have spaces in their writing system. (i.e. Japanese) Meanwhile, in the decoding mechanism this is called silence token and silence is common to all the spoken languages.

Comment (mthrok) [inline examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py:217]: Can you update the return type of the signature and docstring?

Comment (mthrok) [reply examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py:217]: I think `lower` is redundant as the `tokens` passed to the constructor is already lowered.

**Review by mthrok — COMMENTED** [review]
While at it, could you add a note on how KenLM model is trained?

File: examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py [hunk 11]
```diff
@@ -161,26 +163,27 @@
...
-# :py:func:`torchaudio.prototype.ctc_decoder.kenlm_lexicon_decoder`
-# factory function.
-# In addition to the previously mentioned components, it also takes in
-# various beam search decoding parameters and token/word parameters.
+# The decoder can be constructed using the ``kenlm_lexicon_decoder``
+# factory function from ``torchaudio.prototype.ctc_decoder``. In addition
+# to the previously mentioned components, it also takes in various beam
+# search decoding parameters and token/word parameters. The full list of
+# parameters can be found
+# :py:func:`here <torchaudio.prototype.ctc_decoder.kenlm_lexicon_decoder>`.
```
Comment (mthrok) [inline examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py:171]: nit: I think instead of waiting to provided the link to the document at the end end of section (`The full list of parameters ...`), we can link it in the first mention of the function.
(But you might have chosen the current working to work around the formatter warning.) 

```
# The decoder can be constructed using the :py:func:`torchaudio.prototype.ctc_decoder.kenlm_lexicon_decoder`.
# In addition to the previously mentioned components, it also takes in various beam
# search decoding parameters and token/word parameters.
```

File: examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py [hunk 12]
```diff
@@ -222,28 +228,215 @@ def forward(self, emission: torch.Tensor) -> str:
...
+######################################################################
+# beam size token
+# ~~~~~~~~~~~~~~~
+#
+# The ``beam_size_token`` parameter corresponds to the number of tokens to
+# consider for expanding each hypothesis at the decoding step. Exploring a
+# larger number of next possible tokens increases the range of potential
+# hypotheses at the cost of computation. This parameter is especially
+# helpful for lexicon-free decoders, where the number of possible next
+# tokens is not already partially constrained by the lexicon.
```
Comment (mthrok) [inline examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py:358]: >  This parameter is especially helpful for lexicon-free decoders, 
> where the number of possible next tokens is not already
> partially constrained by the lexicon.

Although this is informative and educative, we do not support a lexicon-free decoder (yet?), so maybe not as useful to users?

File: examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py [hunk 13]
```diff
@@ -222,28 +228,215 @@ def forward(self, emission: torch.Tensor) -> str:
...
+    transcript, decode_time = get_transcript_and_time(beam_search_decoder, emission)
+    print(f"beam size token {beam_size_token:<3}: {transcript} ({decode_time:.4f} secs)")
+
+
+######################################################################
+# beam threshold
+# ~~~~~~~~~~~~~~
+#
+# The ``beam_threshold`` parameter is used to prune the stored hypotheses
+# set at each decoding step, removing hypotheses whose scores are greater
```
Comment (mthrok) [inline examples/tutorials/asr_inference_with_ctc_decoder_tutorial.py:383]: Isn't the higher the better?
https://514632-90321822-gh.circle-artifacts.com/0/docs/tutorials/asr_inference_with_ctc_decoder_tutorial.html#beam-size-token

Also, can you also show scores for each example after `nbest`? I was trying to check the score of other examples.

**Review by mthrok — APPROVED** [review]
